OTLP trace gRPC exporter

paivagustavo opened the pull request:
## Quick explanation:

This introduces `otlptrace` package, defining a trace `Client` interface and trace `Exporter` struct that implements the otel sdk exporter interface. **This replaces the `ProtocolDriver` interface and describe an interface specific for the trace exporter.** 

The `otlptracegrpc` adds a client that implements the `otlptrace.Client`. **Most of this was part of the `exporters/otlp/otlpgrpc/driver.go`**.

The `otlp/grpc` folder contains the `grpc` connection logic that will be used by both the metrics and traces `Client`. This still needs to be a module to not add the `grpc` dependency on the `otlp` package. **This is a copy from the existing code in `otlpgrpc/connection.go`**

The `otlptracetest` containts the integration test for the otlp trace `Exporter`, to test it with both `grpc` and `http` clients. **This is a copy from the existing code `exporters/otlp/internal/otlptest`.**

Tests were mainly copied from the existing tests and removed the metrics part of them.

## How users will interact with it
The _sugar_ functions to create and install an pipeline was added directly to the `otlptracegrpc` pacakge. This means that the users only needs to interact with it to create/install a trace pipeline to the sdk.

```go
exp, err := otlptracegrpc.NewExporter(ctx, opts...)
```
or
```go
exp, tp, err := otlptracegrpc.InstallNewPipeline(ctx, opts...)
```


## What comes next?
This is part of the #1827 todo list:
- [x] add gRPC otlp trace exporter
- [ ] add gRPC otlp metrics exporter
- [ ] add HTTP otlp trace exporter
- [ ] add HTTP otlp metrics exporter
- [ ] refactor out the duplicated code (transform, connection, otlpconfig)
- [ ] ~update proto-go dependencies~
- [ ] update examples  and remove old otlp gRPC/HTTP exporter
- [ ] update docs

MrAlias reviewed (commented):
Overall I think the approach show in this PR should be adopted. I think this PR, though large, is properly scoped for the initial PR for a signal in the OTLP and should be marked ready for review so it can be thoroughly reviewed.

I'm assuming there would be follow on PRs to:

- add the HTTP driver for traces
- add a similarly scoped metrics client with gRPC driver
- add an HTTP metrics driver (?)

MrAlias commented:
We should also include a PR to finish the changes that updates the README: https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/otlp/README.md

pellared commented:
It would be good if you rebase your branch. It misses changes introduced here: https://github.com/open-telemetry/opentelemetry-go/commit/d23cc61b93f0fdfe6e0a0fe7869f54d7033aa759

I have one general design comment for the whole `exporters/otlp` directory.

I feel that the the `exporters/otlp` package and `exporters/otlp/otlptrace` have really nothing to do with the `otlp` itselt. It looks just like a helper package to create an exporter (it does not have to be OTLP). Probably it should not be located under `otlp` but just somewhere under `exporters` and also rather as `internal` to reduce the API surface at this stage.

Moreover, I agree with @paivagustavo that these packages should be encapsulated so that the user can jus like this: `exp, err := otlptracegrpc.NewExporter(ctx, opts...)`. It also decouples the implementation details and it makes it possible in the future to not use `otlptrace` without making any breaking changes.

MadVikingGod commented on exporters/otlp/otlptrace/exporter_test.go:
```diff
@@ -0,0 +1,173 @@
+	m.rs = nil
+}
+
+func TestExporterShutdownHonorsTimeout(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	e := otlptrace.NewUnstartedExporter(&stubClient{})
+	if err := e.Start(ctx); err != nil {
+		t.Fatalf("failed to start exporter: %v", err)
+	}
+
+	innerCtx, innerCancel := context.WithTimeout(ctx, time.Microsecond)
+	<-time.After(time.Second)
+	if err := e.Shutdown(innerCtx); err == nil {
+		t.Error("expected context DeadlineExceeded error, got nil")
+	} else if !errors.Is(err, context.DeadlineExceeded) {
+		t.Errorf("expected context DeadlineExceeded error, got %v", err)
+	}
+	innerCancel()
+}
```
Does this test that the Exporter honors shutdown, or that the client does?
For example, if `stubClient` just ignored the `ctx` and spun will it still shutdown?

I know this is a copy, but we might want to move it to the client, or remove it.

paivagustavo replied:
Yes, you are correct. In the past this tested the connection direct, and at the point that the `ProtocolDriver` interface was added, this started to test a stub client, which doesn't really test the actual code. 

I've moved that to the client end-to-end tests, and will be test on both gRPC/HTTP clients.

MadVikingGod commented on exporters/otlp/otlptrace/exporter_test.go:
```diff
@@ -0,0 +1,173 @@
+	innerCancel()
+}
+
+func TestExporterShutdownHonorsCancel(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	e := otlptrace.NewUnstartedExporter(&stubClient{})
+	if err := e.Start(ctx); err != nil {
+		t.Fatalf("failed to start exporter: %v", err)
+	}
+
+	var innerCancel context.CancelFunc
+	ctx, innerCancel = context.WithCancel(ctx)
+	innerCancel()
+	if err := e.Shutdown(ctx); err == nil {
+		t.Error("expected context canceled error, got nil")
+	} else if !errors.Is(err, context.Canceled) {
+		t.Errorf("expected context canceled error, got %v", err)
+	}
+}
```
Same as `TestExporterShutdownHonorsTimeout`

paivagustavo replied:
same as above.

MadVikingGod commented on exporters/otlp/otlptrace/otlptracegrpc/client_test.go:
```diff
@@ -0,0 +1,839 @@
+// Copyright The OpenTelemetry Authors
```
Copied from: otlpgrpc/otlp_integration_test.go.  Just bookkeeping.

MadVikingGod replied:
Sorry, I needed to keep track of this somewhere.  I didn't mean for you to add it.

paivagustavo replied:
Hah, no problem :)

MadVikingGod reviewed (commented):
I like where it's going. I think this would be an acceptable way forward.

I know with the shared `otel/grpc` and `internal` this might not work, but would it be easier to maintain and communicate to users if this was under `exporters/trace/otel`?  I think we pontificated on it a bit in the meeting, so if it's not possible or it makes things messier let's not do it.

paivagustavo commented on exporters/otlp/otlptrace/clients.go:
```diff
@@ -0,0 +1,44 @@
+
+// Client is an interface used by OTLP exporter. It's
+// responsible for connecting to and disconnecting from the collector,
+// and for transforming traces and metrics into wire format and
```
Yes, this was mostly copied from the `ProtocolDriver` which used to export both traces and metrics. Removed the part that it mention "metrics"

Aneurysm9 commented:
> Before the metrics API is stabilized I would prefer to use a module like `x/exporters/otlphttp` which would implement both traces and metrics exporter interfaces. As a user, if I am open to experimentation, I would simply use the module under `x` tree and I could use one exporter as both metrics and trace exporter.
> 
> When we (contributors) feel that the stuff under `x` is stabilized then we simply replace the stable `otlphttp` implementation with the code from `x`. Next, the package `x` can be still used (there is no need to remove it) and it can be then used e.g. to add a log exporter implementation.

I really do not like the idea of using import path to indicate experimental status and requiring users to change all of their imports and references when stability is reached.  Semver has a perfectly fine way of indicating that an API is experimental and unstable, major version 0.  Using a different import path would seem to violate this tenet of the [OTel versioning guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#experimental).

> OpenTelemetry clients MUST NOT be designed in a manner that breaks existing users when a signal transitions from experimental to stable. This would punish users of the release candidate, and hinder adoption.

pellared commented:
> I really do not like the idea of using import path to indicate experimental status and requiring users to change all of their imports and references when stability is reached. Semver has a perfectly fine way of indicating that an API is experimental and unstable, major version 0. \
> Using a different import path would seem to violate this tenet of the [OTel versioning guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#experimental).
> 
> > OpenTelemetry clients MUST NOT be designed in a manner that breaks existing users when a signal transitions from experimental to stable. This would punish users of the release candidate, and hinder adoption.

I see your points. I will think about it a few more times (_maybe I will even edit this comment if totally I change my mind_ 😄 ).

I see the `x` pattern as a way to experiment with new API for modules/packages that already reached `v1+`.

With my proposal I try to achieve the following:
1. Reduce the number of packages so that the users not get lost in the large API/package surface.
2. Still have a way for the maintainers to experiment with a new unstable API.

What other solution do we have? Create a new package each time we want to create a new unstable API? Then the discovery of packages that contain new features would be harder. 

**Loose thoughts** (it is chaotic)

I am aware that the semantic import versioning has its drawbacks: https://peter.bourgon.org/blog/2020/09/14/siv-is-unsound.html
However, one of its main reasons was to reduce the possibility of the diamond dependency problem: https://research.swtch.com/vgo

"Requiring users to change all of their imports" is a general property of Go modules. The users would need to change it anyway if we ever create a new major version. 

Having a different import path for the "stable" and "unstable" packages reduces the possibility that a user would have the diamond problem. If done properly, the diamond problem could only occur for packages that are using unstable imports and the stable one should always work if the "stable" packages follow SemVer.

A broken build is more punishing than just changing the import path. Still, I suggested keeping the source code under `x`. If the unstable API did not have any breaking change then the import path does not have to be changed by any user.

paivagustavo commented:
> Can you please add #1827 to the PR description? I spend a few minutes to find it 😉

Sorry abou that, I've added it to the description along with a todo list for closing that issue. 


> It would be good if you rebase your branch. It misses changes introduced here: [d23cc61](https://github.com/open-telemetry/opentelemetry-go/commit/d23cc61b93f0fdfe6e0a0fe7869f54d7033aa759)

Thanks for catching that, rebased.


About the experimental package, I would say that we keep away of that. The `metrics` is not experimental, it's just unstable and users that are currently using it should be aware of it since it is a 0.x version. This pattern may work for some of the go language packages since they could drop any of them at any time, and that does not match with our current problem about stability of the metrics exporter.

I'm going to mark this PR as ready to review, and will start to get the others PRs moving as well.

paivagustavo commented on exporters/otlp/otlptrace/otlptracegrpc/mock_collector_test.go:
```diff
@@ -0,0 +1,239 @@
+// Copyright The OpenTelemetry Authors
```
Copied and removed metrics part from: exporters/otlp/otlpgrpc/mock_collector_test.go

paivagustavo commented on exporters/otlp/otlptrace/otlptracegrpc/options.go:
```diff
@@ -0,0 +1,174 @@
+// Copyright The OpenTelemetry Authors
```
Copied and removed metrics part from: exporters/otlp/otlpgrpc/options.go

jmacd commented on exporters/otlp/otlptrace/exporter.go:
```diff
@@ -0,0 +1,126 @@
+
+// InstallNewPipeline instantiates a NewExportPipeline with the
+// recommended configuration and registers it globally.
+func InstallNewPipeline(ctx context.Context, client Client) (*Exporter, *tracesdk.TracerProvider, error) {
```
Looks good!

Aneurysm9 reviewed (commented):
I think this needs more godoc and test coverage before I can approve it, but I think this is definitely on the right track.  Let's discuss at the SIG meeting tomorrow how we can sequence the steps you've outlined to get the trace exporters in shape ASAP to enable an RC and then we can worry about metrics.

paivagustavo commented on exporters/otlp/grpc/connection/connection.go:
```diff
@@ -0,0 +1,430 @@
+	maxJitterNanos := int64(0.7 * float64(connReattemptPeriod))
+
+	for {
+		// Otherwise these will be the normal scenarios to enable
```
This was copied from the existing code of the `otlpgrpc` driver. I do have some thoughts if we do need all this reconnection logic if gRPC already handles reconnection. 

I would like to take that to another PR as it related to what this PRs is trying to accomplish.

paivagustavo commented:
I've copied the `transform` and `otlpconfig` packages to the `otlptrace`package. This completely removes the dependency of the `otlp` package from the `otlptrace`, and as a consequence the metrics sdk as well.

This does duplicate a bunch of code, that could be refactored in the future once the old `otlp` drivers are deleted.

Sorry for this big PR.

paivagustavo commented on exporters/otlp/grpc/connection/connection.go:
```diff
@@ -0,0 +1,430 @@
+
+	// these fields are read-only after constructor is finished
+	cfg                  otlpconfig.Config
+	SCfg                 otlpconfig.SignalConfig
```
As this is an internal package again, I'm postponing this for a further refactor.

Aneurysm9 replied:
If it is internal it should be moved to a path that contains "/internal/" somewhere in it.  As it is, this adds to the public API surface so we can't really defer questions about its API.

paivagustavo replied:
Didn't see your reply, but it has already been moved to an internal package.

paivagustavo commented on exporters/otlp/grpc/connection/connection.go:
```diff
@@ -0,0 +1,430 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connection
```
As this is an internal package again, I'm postponing this for a further refactor.

MrAlias commented on exporters/otlp/otlptrace/exporter_test.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+func TestNewExportPipeline(t *testing.T) {
+	testCases := []struct {
+		name string
+	}{
+		{
+			name: "simple pipeline",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			_, tp, err := otlptrace.NewExportPipeline(
+				context.Background(),
+				&noopClient{},
+			)
+
+			assert.NoError(t, err)
+			assert.NotEqual(t, tp, otel.GetTracerProvider())
+
+			_, span := tp.Tracer("otlp test").Start(context.Background(), tc.name)
+			spanCtx := span.SpanContext()
+			assert.Equal(t, true, spanCtx.IsSampled())
+			span.End()
+		})
+	}
```
```suggestion
	_, tp, err := otlptrace.NewExportPipeline(context.Background(), &noopClient{})
	assert.NoError(t, err)
	assert.NotEqual(t, tp, otel.GetTracerProvider())
```

MrAlias commented on exporters/otlp/otlptrace/internal/connection/alignment_test.go:
```diff
@@ -0,0 +1,38 @@
+func TestMain(m *testing.M) {
+	fields := []ottest.FieldOffset{
+		{
+			Name:   "connection.lastConnectErrPtr",
```
```suggestion
			Name:   "Connection.lastConnectErrPtr",
```

MrAlias commented on exporters/otlp/otlptrace/internal/connection/connection.go:
```diff
@@ -0,0 +1,429 @@
+	go c.indefiniteBackgroundConnection()
+
+	// TODO: proper error handling when initializing connections.
+	// 		We can report permanent errors like invalid settings.
```
```suggestion
	// We can report permanent errors like invalid settings.
```

MrAlias commented on exporters/otlp/otlptrace/internal/otlpconfig/envconfig.go:
```diff
@@ -0,0 +1,195 @@
+	"go.opentelemetry.io/otel"
+)
+
+var httpSchemeRegexp = regexp.MustCompile(`(?i)^http://|https://`)
```
```suggestion
var httpSchemeRegexp = regexp.MustCompile(`(?i)^(http://|https://)`)
```

MrAlias commented on exporters/otlp/otlptrace/internal/otlpconfig/tls.go:
```diff
@@ -0,0 +1,57 @@
+	"errors"
+)
+
+const (
+	WeakCertificate = `
+-----BEGIN CERTIFICATE-----
+MIIBhzCCASygAwIBAgIRANHpHgAWeTnLZpTSxCKs0ggwCgYIKoZIzj0EAwIwEjEQ
+MA4GA1UEChMHb3RlbC1nbzAeFw0yMTA0MDExMzU5MDNaFw0yMTA0MDExNDU5MDNa
+MBIxEDAOBgNVBAoTB290ZWwtZ28wWTATBgcqhkjOPQIBBggqhkjOPQMBBwNCAAS9
+nWSkmPCxShxnp43F+PrOtbGV7sNfkbQ/kxzi9Ego0ZJdiXxkmv/C05QFddCW7Y0Z
+sJCLHGogQsYnWJBXUZOVo2MwYTAOBgNVHQ8BAf8EBAMCB4AwEwYDVR0lBAwwCgYI
+KwYBBQUHAwEwDAYDVR0TAQH/BAIwADAsBgNVHREEJTAjgglsb2NhbGhvc3SHEAAA
+AAAAAAAAAAAAAAAAAAGHBH8AAAEwCgYIKoZIzj0EAwIDSQAwRgIhANwZVVKvfvQ/
+1HXsTvgH+xTQswOwSSKYJ1cVHQhqK7ZbAiEAus8NxpTRnp5DiTMuyVmhVNPB+bVH
+Lhnm4N/QDk5rek0=
+-----END CERTIFICATE-----
+`
+	WeakPrivateKey = `
+-----BEGIN PRIVATE KEY-----
+MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgN8HEXiXhvByrJ1zK
+SFT6Y2l2KqDWwWzKf+t4CyWrNKehRANCAAS9nWSkmPCxShxnp43F+PrOtbGV7sNf
+kbQ/kxzi9Ego0ZJdiXxkmv/C05QFddCW7Y0ZsJCLHGogQsYnWJBXUZOV
+-----END PRIVATE KEY-----
+`
+)
+
```
Is there any reason to not move these to the testing files that use them?

paivagustavo replied:
When I initially added these, I thought about reusing that on the http test. Since it remains only used by `options_test`, I've moved these to that file.

MrAlias commented on exporters/otlp/otlptrace/internal/otlptracetest/clienttest.go:
```diff
@@ -0,0 +1,120 @@
+	e := initializeExporter(t, client)
+
+	innerCtx, innerCancel := context.WithTimeout(ctx, time.Microsecond)
+	<-time.After(time.Second)
```
```suggestion
	<-innerCtx.Done()
```

MrAlias commented on exporters/otlp/otlptrace/internal/otlptracetest/clienttest.go:
```diff
@@ -0,0 +1,120 @@
+// Copyright The OpenTelemetry Authors
```
Can we rename this file to `client.go`? It seems redundant to have `test` in both the package name and the file name.

MrAlias commented on exporters/otlp/otlptrace/internal/otlptracetest/clienttest.go:
```diff
@@ -0,0 +1,120 @@
+}
+
+func testClientStopHonorsTimeout(t *testing.T, client otlptrace.Client) {
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
```
What's the point of these "outter" context? They seem superfluous.

paivagustavo replied:
It was supposed to be used on the call to the `NewExporter()`, since I've created a `initializeExporter` to do create an exporter, I've moved this context to that function as well.

MrAlias commented on exporters/otlp/otlptrace/internal/otlptracetest/collector.go:
```diff
@@ -0,0 +1,106 @@
+	tracepb "go.opentelemetry.io/proto/otlp/trace/v1"
+)
+
+// TracesCollector is an interface that mock collectors should implements,
+// so they can be used for the end-to-end testing.
```
```suggestion
// TracesCollector mocks a collector for the end-to-end testing.
```

MrAlias commented on exporters/otlp/otlptrace/internal/otlptracetest/collector.go:
```diff
@@ -0,0 +1,106 @@
+	GetResourceSpans() []*tracepb.ResourceSpans
+}
+
+// SpansStorage stores the spans. Mock collectors could use it to
```
```suggestion
// SpansStorage stores the spans. Mock collectors can use it to
```

MrAlias commented on exporters/otlp/otlptrace/otlptracegrpc/client.go:
```diff
@@ -0,0 +1,111 @@
+	errNoClient = errors.New("no client")
+)
+
+// NewClient creates a new gRPC protocol driver.
```
```suggestion
// NewClient creates a new gRPC trace client.
```

MrAlias commented on exporters/otlp/otlptrace/otlptracegrpc/client.go:
```diff
@@ -0,0 +1,111 @@
+	}
+}
+
+// Start implements otlptrace.Client. It establishes a connection
+// to the collector.
```
```suggestion
// Start establishes a connection to the collector.
```

MrAlias commented on exporters/otlp/otlptrace/otlptracegrpc/client.go:
```diff
@@ -0,0 +1,111 @@
+	return c.connection.StartConnection(ctx)
+}
+
+// Stop implements otlptrace.Client. It shuts down the connection
+// to the collector.
```
```suggestion
// Stop shuts down the connection to the collector.
```

MrAlias commented on exporters/otlp/otlptrace/otlptracegrpc/client.go:
```diff
@@ -0,0 +1,111 @@
+	return c.connection.Shutdown(ctx)
+}
+
+func (c *client) UploadTraces(ctx context.Context, protoSpans []*tracepb.ResourceSpans) error {
```
missing comment.

MrAlias commented on exporters/otlp/otlptrace/otlptracegrpc/options.go:
```diff
@@ -0,0 +1,175 @@
+	applyGRPCOption(*otlpconfig.Config)
+}
+
+type RetrySettings otlpconfig.RetrySettings
```
missing comment.

MrAlias reviewed (approved):
:1st_place_medal:

MrAlias commented:
For the test failures, `TraceStateFromKeyValues` was moved to the `oteltest` package for testing purposes. I think the only place it is being used here is in the test packages so this should be a simple fix of changing the package it is imported from.

pellared commented:
I am still not sure about the design of `otlptrace` module.

I feel this is a smell:
```
func NewExporter(ctx context.Context, opts ...Option) (*otlptrace.Exporter, error) {
```

The `otlptrace` looks for me more like a helper module which only provides a common exporter API implementation. I would rather make the package internal and wrap the type. 

Going further with current designl, the `otlptracehttp` would have the same coupling to `otlptrace` and it will be e.g. impossible to change the `otlptrace.Exporter` without possibly affecting both modules.

paivagustavo commented:
> Should the user ever need to use the otlptrace module directly?

They way it is designed, users and vendors _may_ implement their own client. One of the reasons for such custom client would be https://github.com/open-telemetry/opentelemetry-go/issues/1819#issuecomment-829585683.

>It will be hard to make a gRPC/HTTP specific change in the otlptrace.Exporter without without breaking Liskov Substitution Principle.

The Exporter is only supposed to implement the `tracesdk.SpanExporter` interface and nothing else. Any protocol-specific code should live in the specific client package.

I understand your reasoning about the existence of a `otlptrace` package. It is indeed a "helper" package, that it depends on others packages to work. I like the idea of keeping that "public",i.e. not internal, and letting users extend if needed. 

We can go either direction, and it is easy to move this package to an `internal` folder if there is a consensus about that, we should probably discuss more on this week SIG meeting.

pellared commented:
> > Should the user ever need to use the otlptrace module directly?
> 
> They way it is designed, users and vendors _may_ implement their own client. One of the reasons for such custom client would be [#1819 (comment)](https://github.com/open-telemetry/opentelemetry-go/issues/1819#issuecomment-829585683).
> 
> > It will be hard to make a gRPC/HTTP specific change in the otlptrace.Exporter without without breaking Liskov Substitution Principle.
> 
> The Exporter is only supposed to implement the `tracesdk.SpanExporter` interface and nothing else. Any protocol-specific code should live in the specific client package.
> 
> I understand your reasoning about the existence of a `otlptrace` package. It is indeed a "helper" package, that it depends on others packages to work. I like the idea of keeping that "public",i.e. not internal, and letting users extend if needed.
> 
> We can go either direction, and it is easy to move this package to an `internal` folder if there is a consensus about that, we should probably discuss more on this week SIG meeting.

Even if we want to have `otlptrace` public I would consider wrapping it just in case. I do not have a very strong opinion, I just try to be very careful about the public API 😋

Aneurysm9 commented:
> It will be hard to make a gRPC/HTTP specific change in the `otlptrace.Exporter` without without breaking Liskov Substitution Principle.

Go doesn't have subclassing, so I'm not sure that the LSP is applicable.  If you want to think in terms of classical inheritance, I would view `otlptrace` as providing an abstract class that needs to be reified by including a client implementation.  That implementation can come from `otlptracegrpc` or `otlptracehttp` or some other user-provided package, but since we use composition and not inheritance the relationship is inverted and the `otlptrace.Exporter` has its missing implementation provided by composing a `otlptrace.Client` implementation rather than having other packages inherit common functionality from it.

pellared commented:
> Go doesn't have subclassing, so I'm not sure that the LSP is applicable. If you want to think in terms of classical inheritance, I would view `otlptrace` as providing an abstract class that needs to be reified by including a client implementation. That implementation can come from `otlptracegrpc` or `otlptracehttp` or some other user-provided package, but since we use composition and not inheritance the relationship is inverted and the `otlptrace.Exporter` has its missing implementation provided by composing a `otlptrace.Client` implementation rather than having other packages inherit common functionality from it.

```
func NewExporter(ctx context.Context, opts ...Option) (*otlptrace.Exporter, error) {
```

My concern is that this function returns a concrete struct. Not an "abstract class". 

I am not against using `otlptrace.Exporter` for sake of easier implementation. I think it should not be used as a "returned value" in the function signature. It tightly couples the `NewExporter` functions in `otlptracegrpc` and `otlptracegrpc` to `otlptrace.Exporter`.

pellared commented:
_Writing from mobile_ I remember that some Options where wrapped in otlp folder to mitigate similar issue

MadVikingGod commented:
Not a blocker because this is internal.

I noticed that the grpc connection struct was moved into the otlptrace package.  Is the plan to have this mostly copied when we make the otlpmetric?  If we move it back to otlp, or have copies I think either will work.

pellared reviewed (approved):
During SIG meeting @Aneurysm9 has explained why we are OK using the same exporter struct for both modules.
